### ISSUE TITLE
rtp streaming: after a few playback, the receiver's buffering shows 0.0, i.e., cannot receive any data

### ISSUE THREAD

**Issue by mikej1688** [caprica/vlcj#994]
I tested how the rtp streaming worked by using vlc4.0.0/vlcj5.0.0. The test used two windows 10 boxes in a private network. Used the box A as a rtp streaming data sender to stream video files, and the box B as the receiver. The network communication happened within an intranet.
The test was like this: initially the box A sent out a video clip, and the receiver received it and played fine, although the quality was poor with a lot of errors shown on the receiver side as follows,
[1ec6a928] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 9, expected 8) for PID 32
[1ec6a928] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 9, expected 8) for PID 17
[1ec6a928] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 14, expected 13) for PID 0
[1ec6a928] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 14, expected 13) for PID 32
[1ec6a928] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 14, expected 13) for PID 17

later on, when I kept playing more video files, it was found that once such errors (still on the receiver side)

[1c62dc50] main input error: Your input can't be opened
[1c62dc50] main input error: VLC is unable to open the MRL 'rtp://@:2796'. Check the log for details.

appeared, the receiver no longer received any data from the sender. I guessed it was probably too early for the receiver to listen to the rtp port?  like this,

rtp://@:2796

However, when I tested the vlc media player, the media player can be listening there, i.e., in a state of waiting, even before the stream sender began to send the streaming data. Can I set a wait period of time for the listener to wait for? if yes, what's the config param?

On the receiver side, the EmbeddedMediaPlayer implemented MediaPlayerEventListener.buffering(MediaPlayer mediaPlayer, float newCache) method. After the above error occurred, the buffering method shown the received data by the buffer was always 0.0. 
In this case, I removed the existing MediaPlayerEventListener and added a new listener to the mediaPlayer, but the issue remained.

Here are the typical sample code.

The RTP sender side:

```
public class RTPStreamServer {
	private static final String SOUT_KEEP = ":sout-keep";
	private static final String TTL_1 = ":ttl=1";
	private static final String SOUT_ALL = ":sout-all";
	private static final String NO_TS_TRUST_PCR = ":no-ts-trust-pcr";
	private static final String TS_SEEK_PERCENT = ":ts-seek-percent";
			
	private static RTPStreamServer instance = null;

	private MediaPlayerFactory mediaPlayerFactory = null;
	private MediaPlayer mediaPlayer = null;

	public static RTPStreamServer getInstance() {
		if (instance == null) {
			synchronized (RTPStreamServer.class) {
				if (instance == null) {
					instance = new RTPStreamServer();
				}
			}
		}
		return instance;
	}

	private RTPStreamServer() {
		mediaPlayerFactory = new MediaPlayerFactory();
		mediaPlayer = mediaPlayerFactory.mediaPlayers().newMediaPlayer();
	}

	public void streamData(String protocolType, String mediaUrl, Endpoint targetEndpoint) throws InterruptedException {
		options = getStreamOptionString(mediaUrl, targetEndpoint, protocolType);
		if (mediaPlayer != null && mediaPlayer.status().isPlaying()) {
			mediaPlayer.controls().stop();
		}
		mediaPlayer.media().play(
				mediaUrl, 
                                options, 
				NO_TS_TRUST_PCR,
				TS_SEEK_PERCENT,
				SOUT_ALL, 
				TTL_1,
				SOUT_KEEP);
	}

	private String getStreamOptionString(String mediaUrl, Endpoint targetEndpoint, String protocolType) {
		StringBuilder sb = new StringBuilder(150);

		sb.append(":sout=#" + protocolType.toLowerCase() + "{dst=");
		sb.append(targetEndpoint.getLocalAddress());
		sb.append(",port=");
		sb.append(targetEndpoint.getLocalDataPort());
		sb.append(",mux=ts}");
		return sb.toString();
	}
     
       ...
}

```
The RTP receiver side:

```
public class RTPStreamClient {
	private static final String SOUT_KEEP = ":sout-keep";
	private static final String SOUT_ALL = ":sout-all";
	private static final String NO_TS_TRUST_PCR = ":no-ts-trust-pcr";
	private static final String TS_SEEK_PERCENT = ":ts-seek-percent";

	public void initializeMediaPlayer() {
		NativeLibrary.addSearchPath(RuntimeUtil.getLibVlcLibraryName(), getVLCNativeFolder());
		LibC.INSTANCE._putenv("VLC_PLUGIN_PATH=" + getVLCNativeFolder());
		mediaPlayerFactory = new MediaPlayerFactory();
		mediaPlayer = mediaPlayerFactory.mediaPlayers().newEmbeddedMediaPlayer();
		mediaPlayer.videoSurface().set(mediaPlayerFactory.videoSurfaces().newVideoSurface(canvas));
		mediaPlayer.controls().setRepeat(false);
		listener = new RTPMediaPlayerEventAdapter();
		mediaPlayer.events().addMediaPlayerEventListener(listener);
		mediaPlayerInitialized = true;
	}

        /**
          * @param url the listening url, like "rtp://@:2796"
          */
	public void lazyStartPlayer(String url)  throws Exception {                 
		if (!mediaPlayerInitialized) {
			initializeMediaPlayer();
		}
		if (mediaPlayer != null) {
			mediaPlayer.controls().stop();
			Thread.sleep(2000L);
			mediaPlayer.media().play(
				url, 
				NO_TS_TRUST_PCR, 
				TS_SEEK_PERCENT, 
				SOUT_ALL, 
				SOUT_KEEP); 
		}
	}
}
```

```
class RTPMediaPlayerEventAdapter extends MediaPlayerEventAdapter {

	@Override
	public void buffering(MediaPlayer mediaPlayer, float newCache) {
		System.out.println("Buffering " + newCache);
	}
}
```

**Comment by caprica** [comment]
If you use three back-ticks before and after your code blocks it won't be so difficult to read like it is here.

**Comment by caprica** [comment]
This is difficult to offer any suggestions on.

VLC itself does not use LibVLC, so it will have certain advantages over LibVLC applications - perhaps like network resiliance.

You should also try and make things more event-based, I notice your sleep because stop() is not synchronous - and you don't have even that sleep in your streamData method after a stop. There's also a nicer pattern for thread-safe one-time singleton initialisation (IMO) using a static inner class. These are side issues of course, but when things aren't quite working right I like to make sure everything else is solid.

Apart from that I can only suggest creating a MediaPlayerFactory with "-vvv" arguments and checking the resulting VLC native logs.

It's unlikely this sort of problem can be solved on the Java side to be honest.

**Comment by mikej1688** [comment]
Thanks for the suggestion. 
yes, I also refactored the code like, 
```
      public void streamData(String protocolType, String mediaUrl, Endpoint targetEndpoint) throws InterruptedException {
		options = getStreamOptionString(mediaUrl, targetEndpoint, protocolType);
		if (mediaPlayer != null && mediaPlayer.status().isPlaying()) {
			mediaPlayer.controls().stop();
		}
		while(mediaPlayer.status().isPlaying()) {
			Thread.sleep(100);
		}
		mediaPlayer.media().play(
				mediaUrl, 
                                options, 
				NO_TS_TRUST_PCR,
				TS_SEEK_PERCENT,
				SOUT_ALL, 
				TTL_1,
				SOUT_KEEP);
	}
```
The situation remained.
I tried to use HTTP/TCP streaming over network. In this case, the quality of playback on the receiver side was pretty good,  compared with the case using RTP/UDP. It shown much fast buffering rate(why?). However, it crashed the whole streaming server when playing the second video file. The issue seemed to only happen for the latest vlc4.0.0 nightly build. For the old versions of vlc/vlcj, it seemed fine. But I still prefer to use RTP/UDP for audio/video streaming. 
I also attempted playing with RTSP/RTP streaming. In principle, it is still a TCP used for meta data control with data transfer using RTP, right?
An interesting thing is that in StreamRtsp.java, it said, "The client specifies an MRL like, rtsp://@127.0.0.1:5555/demo", if the stream option string is defined by 

String options = formatRtspStream("127.0.0.1", 5555, "demo");

It's similar to the StreamHttp server case. If I run the StreamRtsp as the streaming server on the box A, and from the same box A, also run the streaming receiver, it worked fine. However, if the receiver was on the box B, the streaming was failed. Not quite sure what happened. Probably it was the issue of Windows Defender Firewall blockage. But I'd like to know what difference between HTTP streaming and RTSP streaming is?  Would I prefer RTSP over HTTP?

**Comment by caprica** [comment]
Again, it's difficult to say.

A potential red flag that jumps out is you say you played the next video file on the server before the last video file was finished - how do you do this? Because you can't stream two different videos concurrently using the same network port. So you should wait until one video is finished before playing the next, or maybe consider using a MediaListPlayer instead.

**Comment by caprica** [comment]
And obviously, if you're using nightly builds of VLC there's a quite a bit of risk there.

Is your server piece Linux or Windows? I've heard reports of stuff not working nicely with RTSP/RTP on Windows before.

**Comment by mikej1688** [comment]
Both the server box and the receiver box are Windows 10. 
It was like the situation as people watched Youtube clips - kept changing from one to another, without waiting for the existing one to be finished.  Use of MediaListPlayer would require a pre-compiled list of videos, which might not fit the habit of users.
Here was what was done: before changing to stream the new video, stopped the existing streaming by stopping the media player (but not released). So I hoped to avoid the potential concurrently streaming. 
On the server side, I used a generic MediaPlayer object for streaming video. At the same time, used an EmbeddedMediaPlayer object to play the same video locally. It implied there were 2 media players on one jvm, although run by different threads. Would it cause a problem?  I also tried to use the vlc's option of deplicate ( i.e., streaming out & playing the same video locally), again it didn't work using vlc 4.0.0.
